### DISCUSSION TITLE
Assistance with occasional crash on GTAV load

### DISCUSSION THREAD

**Discussion by nomakewan** [scripthookvdotnet/scripthookvdotnet#1349] [Q&A] [answered]
I have two users who've reported that they cannot get GTAV to load while our mod is installed, and that it crashes to desktop during the "Entering Story Mode" screen, well before the game loads in. I was unable to help them since I didn't experience the same issue on my end. But just today I finally ran into that problem. The game was running fine earlier this morning, but this afternoon when I went back to do more dev work, suddenly I was getting a hard crash while loading into the game.

When this happens, the following .NET Runtime error is left in the Event Viewer:

```
Application: GTA5.exe
Framework Version: v4.0.30319
Description: The process was terminated due to an unhandled exception.
Exception Info: System.AccessViolationException
   at SHVDN.NativeFunc+NativeTaskPtrArgs.Run()
   at SHVDN.ScriptDomain.ExecuteTaskWithGameThreadTlsContext(SHVDN.IScriptTask)
   at SHVDN.NativeFunc.Invoke(UInt64, UInt64*, Int32)
   at GTA.Game.get_IsLoading()
   at BackToTheFutureV.Main.Main_Tick(System.Object, System.EventArgs)
   at SHVDN.Script.DoTick()
   at SHVDN.Script.MainLoop()
   at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
   at System.Threading.ThreadHelper.ThreadStart()
```

At least according to this, there seems to be an access violation related to the following piece of code in BTTFV's main script:

```cs
            if (Game.IsLoading || FusionUtils.FirstTick)
            {
                LoadingPrompt.Show("Loading BTTFV");
                Screen.FadeOut(0);
                return;
            }
```
However, it's also possible that this is a red herring and that there is a more fundamental problem at play here. As I currently don't have the ability to debug the game while it's running and catch the crash in the act, this is all the information I have to go on for now.

Any clues? Or anything I can do to try to dig deeper?

Thanks as always!

**Comment by kagikn** [comment]
I'm assuming the two users used SHVDN v3.6.0 at the time they had their game get crashed, but are you sure which version(s) of SHVDN the two users used? So I can diagnose more quickly. If they used builds after v3.6.0, it might suggest I have messed up something in `ScriptDomain`.

**Reply by nomakewan** [reply]
One of the users is the same as me--using the current bleeding-edge dev build of SHVDN, and the current bleeding-edge dev build of both BTTFV and Fusion Library. For them, reverting to SHVDN 3.5.1 and BTTFV 2.3.2 resolves the crash.

The other user is an outlier as their game crashed with SHVDN 3.5.1 and the older 2.3.2 version of BTTFV. They had no mods other than BTTFV 2.3.2 installed, and a fresh copy of GTAV 1.0.3028.0 from Steam. Since their crash was occurring in the same place as us I thought to mention it. I will ask them to check their Event Viewer to see if they have the same .NET Runtime error.

**Reply by kagikn** [reply]
Understood. Looks like I need to investigate `ScriptDomain` to see if it has trouble in some race conditions.

**Reply by kagikn** [reply, marked as answer]
Pushed a3cf789 for an potential solution. You're going have to see if you (and your people) cannot reproduce the issue. I believe I can't reproduce the issue in my PC, which has i5-4670K and an HDD where my GTA V is installed, for shit because it is not too powerful to reproduce the issue for once.

**Reply by nomakewan** [reply]
In my testing, I no longer run into the "GTAV has stopped responding..." prompt. Whether it solves the problem with my fellow dev user or not I won't know until tomorrow, but I've passed him the files to test with. The third user really is an outlier and I've confirmed with him that his problem is unrelated to ours (his crash leaves nothing in the Event Viewer, and appears to be a file I/O issue).

On my end however I can still cause the .NET Runtime access violation crash (identical to the log I posted) with [a3cf789](https://github.com/scripthookvdotnet/scripthookvdotnet/commit/a3cf789db7bbe308bb083b8273e9eb211ab267a2) if I have OBS running with Game Capture enabled and set to GTA5.exe _before_ launching GTAV. This crash does not occur with SHVDN 3.6.0 or earlier, and appears to have been introduced in [dde6861](https://github.com/scripthookvdotnet/scripthookvdotnet/commit/dde686112cae9054dc75f8876201041eb2dd0957). Builds from main prior to that commit do not exhibit the same crash.

Note that since our dev code now relies on bleeding-edge components (such as GTA.Chrono) our scripts are not actually running correctly on these older SHVDN builds (on commit 1f001b2 and earlier the game loads fine but the console displays errors for BTTFV and FusionLibrary since GameClockDateTime doesn't exist). But regardless, builds after that commit will crash 100% of the time during the game load sequence before entering Story Mode.

Hopefully that helps narrow things down. If there is anything else I can do on my end to assist please let me know.

EDIT: My fellow dev user has reported back and stated that indeed, commit a3cf789 fixes his CTD on first load.

**Reply by kagikn** [reply]
Glad to hear that the commit a3cf789 did do something that prevents from some potential game crash for you and your fellow dev user.

Unfortunately, I haven't reproduced the .NET Runtime access violation crash being thrown that you mentioned. Maybe I would end up finding the cause if you post a dmp file generated when the game crashes. You can find crash dumps the game or RGL creates when the game crashes at `%LOCALAPPDATA%\Rockstar Games\GTAV\CrashLogs`. I can find the line of one of SHVDN source code files at which an exception thrown with an appropriate dmp file the game or RGL creates. Please let me know the commit at which you have your game crashes with the SHVDN build as well when you post a dmp file, so I can find the appropriate pdb file quickly!

**Reply by nomakewan** [reply]
Unfortunately it does not appear that any dumps are generated from this particular crash. Only the Application Hang resulted in a dump, which I have attached. However, I believe that this is the bug you already fixed in commit a3cf789.

The hang in this log was using commit 0676328, I believe.

[log.zip](https://github.com/scripthookvdotnet/scripthookvdotnet/files/13376680/f3b45945-1ac6-4f7b-9167-00d9de804e38.zip)

EDIT: And now I can't get the crash to happen again anyway. If OBS is running and Game Capture is active and targeting GTA5.exe it will cause the game to freeze up (but not crash) during the loading sequence if focused in borderless windowed DX11 mode, but unfocusing it resolves that within a few seconds and then the game loads normally. Exiting OBS allows for a normal load regardless.

Several hours ago, I could get it to happen 100% of the time without fail. The only difference between then and now is time, so I'm honestly at a loss to explain what's going on.

I'm going to go ahead and say that for now, it seems like this is "solved." If a crash happens again and leaves a log, I'll come back to it. But I agree that without a decent log or a way to reliably replicate the issue, there's not much that can be done.

**Reply by kagikn** [reply]
I appreciate your report about the occasional crash that could happen when SHVDN calls an game function, now I understand how cache coherence and data race matter a bit more. I wouldn't have realized the issue without a report like yours.

Btw, I'm afraid to say that I couldn't debug the dmp file you uploaded with WinDbg for mismatching. dbg files vary a bit by the time corresponding source code built iirc. [ChkMatch](https://debuginfo.com/tools/chkmatch.html) didn't help too much either. Reproducing a game crash with binary files built by a workflow run on GitHub actions or providing pdbs and corresponding binary files would help me debugging more than providing pdbs but no info about binary files.

Man, didn't think the crash related to OBS and `System.AccessViolationException` bothers us that much 😭

**Reply by nomakewan** [reply]
Ah, right, that makes total sense since I didn't use the workflow run build back then. I compiled locally.

If it happens again and leaves a dump, I'll be sure to include all the compiled SHVDN content along with the dump to be extra sure.
